sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -0,0 +1,623 @@
+
+      case DateType =>
+        val dateLiteral = litDataType match {
+          case StringType =>
+            DateTimeUtils.stringToDate(litValue.asInstanceOf[UTF8String])
+              .getOrElse(0).toString
+          case _ => litValue.toString
+        }
+        BigDecimal(java.lang.Long.valueOf(dateLiteral))
+
+      case TimestampType =>
+        val tsLiteral = litDataType match {
+          case StringType =>
+            DateTimeUtils.stringToTimestamp(litValue.asInstanceOf[UTF8String])
+                .getOrElse(0).toString
+          case _ => litValue.toString
+        }
+        BigDecimal(java.lang.Long.valueOf(tsLiteral))
+    }
+  }
+
+  /**
+   * Returns a percentage of rows meeting an equality (=) expression.
+   * This method evaluates the equality predicate for all data types.
+   *
+   * @param attrRef an AttributeReference (or a column)
+   * @param literal a literal value (or constant)
+   * @param update a boolean flag to specify if we need to update ColumnStat of a given column
+   *               for subsequent conditions
+   * @return a doube value to show the percentage of rows meeting a given condition
+   */
+  def evaluateEqualTo(
+      attrRef: AttributeReference,
+      literal: Literal,
+      update: Boolean)
+    : Option[Double] = {
+
+    val aColStat = mutableColStats(attrRef.exprId)
+    val ndv = aColStat.distinctCount
+
+    // decide if the value is in [min, max] of the column.
+    // We currently don't store min/max for binary/string type.
+    // Hence, we assume it is in boundary for binary/string type.
+    val inBoundary: Boolean = attrRef.dataType match {
+      case _: NumericType | DateType | TimestampType =>
+        val statsRange =
+          Range(aColStat.min, aColStat.max, attrRef.dataType).asInstanceOf[NumericRange]
+        val lit = numericLiteralToBigDecimal(attrRef.dataType, literal.value, literal.dataType)
+        (lit >= statsRange.min) && (lit <= statsRange.max)
+
+      case _ => true  /** for String/Binary type */
+    }
+
+    if (inBoundary) {
+
+      if (update) {
+        // We update ColumnStat structure after apply this equality predicate.
+        // Set distinctCount to 1.  Set nullCount to 0.
+        val newStats = attrRef.dataType match {
+          case _: NumericType =>
+            val newValue = Some(literal.value)
+            aColStat.copy(distinctCount = 1, min = newValue,
+              max = newValue, nullCount = 0)
+          case DateType =>
+            val dateValue = literal.dataType match {
+              case StringType =>
+                Some(Date.valueOf(literal.value.toString))
+              case _ => Some(DateTimeUtils.toJavaDate(literal.value.toString.toInt))
+            }
+            aColStat.copy(distinctCount = 1, min = dateValue,
+              max = dateValue, nullCount = 0)
+          case TimestampType =>
+            val tsValue = literal.dataType match {
+              case StringType =>

```

Comment:
See `TypeCoercion#L324-L334`, we will convert to string and do string comparison.